### PR TITLE
Fix batch transform issue for tabular predictor with multiple partitions

### PR DESCRIPTION
**Pull request by tonyhoo**
### Description:
This PR fixes the issue where batch transform jobs fail due to column misalignment when the input CSV file is partitioned into multiple records. The problem arises because headers from different partitions are not handled properly, leading to misaligned columns and prediction failures during inference.

**Changes:**
- Added logic to align columns across partitions by ensuring headers are managed correctly.
- Introduced `_read_with_fallback` and `_align_columns` helper functions to handle column alignment.
- Updated `transform_fn` in `tabular_serve.py` to use these helper functions.

**Limitations:**
- This fix currently only works for the tabular predictor. Support for multimodal and timeseries predictors depends on the implementation of `original_features`, which can be tracked in [issue #4477](https://github.com/autogluon/autogluon/issues/4477).

**Steps to Reproduce:**
The following script can be used to reproduce the issue:

```python
from autogluon.cloud import TabularCloudPredictor
import pandas as pd

# Load datasets
train_data = pd.read_csv("https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv")
test_data = pd.read_csv("https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv")
test_data.drop(columns=['class'], inplace=True)

# Cloud Predictor Arguments
predictor_init_args = {"label": "class"}  
predictor_fit_args = {"train_data": train_data, "time_limit": 60}  

# Initialize Cloud Predictor and Fit
cloud_predictor = TabularCloudPredictor(cloud_output_path='tonyhu-autogluon')
cloud_predictor.fit(predictor_init_args=predictor_init_args, predictor_fit_args=predictor_fit_args)

# Batch Inference with small max_payload to force multiple partitions
result = cloud_predictor.predict(test_data, backend_kwargs={"transformer_kwargs": {"max_payload": 1}})
```

**Expected Behavior:**
The batch transform job should handle multiple partitions correctly, aligning columns across the partitions and ignoring or managing headers if present in individual partitions.

**Observed Behavior:**
The job fails with the following error logs:
```
Bad HTTP status received from algorithm: 500
invalid literal for int() with base 10: '0.1': Error while type casting for column 'capital-loss'
```

Logs show that the columns are misaligned for certain partitions:
```
test_columns: [' 11th', ' Machine-op-inspct', ' Male', ' Never-married', ' Other-relative', ' Private', ' United-States', ' White', '0', '0.1', '207443', '50', '62', '7']
2024-09-13T21:56:19,062 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - train_columns: ['age', 'capital-gain', 'capital-loss', 'education', 'education-num', 'fnlwgt', 'hours-per-week', 'marital-status', 'native-country', 'occupation', 'race', 'relationship', 'sex', 'workclass']
```

**Environment:**
- `autogluon==1.1.0`
- Running batch transform in SageMaker with `MultiRecord` strategy.
- `MaxPayloadInMB=1` is set to ensure multiple partitions.

**Additional Information:**
The issue seems to be that AutoGluon Cloud is not handling the headers properly when dealing with batch transform partitioned records. In a multi-partition job, not all batches will have the header/column, which is causing the column misalignment.

**Note:**
This fix currently only works for the tabular predictor. Support for multimodal and timeseries predictors depends on the implementation of `original_features`, which can be tracked in [issue #4477](https://github.com/autogluon/autogluon/issues/4477).

### CODE REVIEW

File: src/autogluon/cloud/backend/sagemaker_backend.py [hunk 1]
```diff
@@ -1104,15 +1112,6 @@ def _predict_real_time(self, test_data, accept, split_pred_proba=True, inference
...
     def _upload_batch_predict_data(self, test_data, bucket, key_prefix):
-        # If a directory of images, upload directly
-        if isinstance(test_data, str) and not os.path.isdir(test_data):
-            # either a file to a dataframe, or a file to an image
-            if is_image_file(test_data):
-                logger.warning(
-                    "Are you sure you want to do batch inference on a single image? You might want to try `deploy()` and `predict_real_time()` instead"
-                )
-            else:
-                test_data = load_pd.load(test_data)
```
Comment (suzhoum) [inline src/autogluon/cloud/backend/sagemaker_backend.py:1115]: Will it cause problem if images are uploaded here?

File: src/autogluon/cloud/backend/sagemaker_backend.py [hunk 2]
```diff
@@ -1179,6 +1179,27 @@ def _predict(
             test_data = convert_image_path_to_encoded_bytes_in_dataframe(
                 dataframe=test_data, image_column=test_data_image_column
             )
+
+        # If a directory of images, upload directly
+        if isinstance(test_data, str) and not os.path.isdir(test_data):
+            # either a file to a dataframe, or a file to an image
+            if is_image_file(test_data):
+                raise ValueError("Image file is not supported for batch inference")
```
Comment (suzhoum) [inline src/autogluon/cloud/backend/sagemaker_backend.py]: Maybe reword to "Single image file"? And suggest to try `.deploy()`

File: src/autogluon/cloud/backend/sagemaker_backend.py [hunk 3]
```diff
@@ -1268,8 +1289,8 @@ def _predict(
             pred = results
             if split_pred_proba:
                 pred, pred_proba = split_pred_and_pred_proba(results)
-        if not persist:
-            os.remove(results_path)
+            if not persist:
+                os.remove(results_path)
```
Comment (suzhoum) [inline src/autogluon/cloud/backend/sagemaker_backend.py:1293]: Good catch!

File: src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py [hunk 4]
```diff
@@ -40,60 +40,50 @@ def model_fn(model_dir):
...
             break
-    # save image column bytes to disk and update the column with saved path
+
     if image_column is not None:
         print(f"Detected image column {image_column}")
         data[image_column] = [_save_image_and_update_dataframe_column(bytes) for bytes in data[image_column]]
 
+    # Ensure inference_kwargs is a dictionary right before use
+    if inference_kwargs is None:
+        inference_kwargs = {}
```
Comment (suzhoum) [inline src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py:84]: The check seems to be redundant?

File: src/autogluon/cloud/backend/sagemaker_backend.py (hunk 1 above)
Comment (tonyhoo) [reply src/autogluon/cloud/backend/sagemaker_backend.py:1115]: These logics have been copied over to line #L1184 - #L1192

File: src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py (hunk 4 above)
Comment (tonyhoo) [reply src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py:84]: No, this code are moved from previous #L64 where to make sure if `inference_kwargs` are explicitly set up to None in payload, we will still have the dict format

File: src/autogluon/cloud/backend/sagemaker_backend.py (hunk 1 above)
Comment (suzhoum) [reply src/autogluon/cloud/backend/sagemaker_backend.py:1115]: I figured. Now after I read it again, I think moving this to `_predict()` makes sense.

File: src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py (hunk 4 above)
Comment (suzhoum) [reply src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py:84]: I think you updated line #65 to make the default value as `{}`, so `inference_kwargs` should not be `None`. But this is a good safety nest.

**Review by suzhoum — APPROVED** [review]
Thanks for the fix!

File: src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py (hunk 4 above)
Comment (tonyhoo) [reply src/autogluon/cloud/scripts/sagemaker_scripts/tabular_serve.py:84]: The default will be set to `{}` only if `inference_kwargs ` not available from `payload`, it will be set to `None` if 
```
payload = {
    "inference_kwargs": None
}
```
